[lldb] Add bidirectional packetLog to gdbclientutils.py

dsandersllvm opened the pull request:
While debugging the tests for #155000 I found it helpful to have both sides
of the simulated gdb-rsp traffic rather than just the responses so I've extended
the packetLog in MockGDBServerResponder to record traffic in both directions.
Tests have been updated accordingly

DavidSpickett commented on lldb/packages/Python/lldbsuite/test/gdbclientutils.py:
```diff
@@ -86,8 +88,8 @@ class MockGDBServerResponder:
     """
 
-    registerCount = 40
-    packetLog = None
```
This predates you, but I wonder why we even have a class level attribute for this? Maybe I am misunderstanding some aspect of Python because it seems to me that the packet log should never be None. Empty list, sure, but not None.

DavidSpickett replied:
Assuming it passes testing, https://github.com/llvm/llvm-project/pull/162453 will remove this line.

DavidSpickett commented on lldb/packages/Python/lldbsuite/test/gdbclientutils.py:
```diff
@@ -306,7 +309,8 @@ def haltReason(self):
 
-    def qXferRead(self, obj, annex, offset, length):
+    def qXferRead(self, obj: str, annex: str, offset: int,
+                  length: int) -> tuple[str | None, bool]:
```
In future please separate annotations of existing types into their own PR. If you're changing the type in the process of doing something then it's fine to include in the same PR.

DavidSpickett commented on lldb/packages/Python/lldbsuite/test/gdbclientutils.py:
```diff
@@ -374,15 +378,17 @@ class UnexpectedPacketException(Exception):
     """
 
-    def get_connect_address(self):
+    @abstractmethod
```
Good idea, but separate PR please.

dsandersllvm committed:
[lldb] Add type hints to gdbclientutils.py and fix issues

Everything in this should be python 3.9. The docs say the minimum is 3.8
but there's existing code in this suite that needs 3.9 so I think 3.9 is
ok.

Issues:
qEcho() is passed an argument by the callers that the function didn't have
Several functions in the base class would silently do nothing if not
overriden. These now use @abstractmethod to require overrides
sendall() had inconsistent return types between overrides

dsandersllvm committed:
[lldb] Make MockGDBServerResponder return consistent types

It used return either:
* str
* The RESPONSE_DISCONNECT class
* The RESPONSE_NONE class
* A list of the above three
it now returns:
* A list of str or SpecialResponse instances

dsandersllvm committed:
[lldb] Renamed assertPacketLogContains to assertPacketLogReceived

The docstring is currently wrong about the packet log containing both sent
and received packets. This will be corrected in the next commit at which
point this function will be filtering to the received packets for
compatibility

dsandersllvm committed:
fixup: Merge the MockGDBServer packetLog into the existing MockGDBServerResponder's packetLog

dsandersllvm commented:
Updated to be a single packetLog in MockGDBServerResponder containing both sent and received packets and the tests have been updated so that they still only check the received packets.

Along the way I noticed that the RESPONSE_DISCONNECT and RESPONSE_NONE classes were being returned by responses. It's not technically wrong in python but mixing instances and classes in a list is a bit odd and it was making type hints tricky to write so these are now instances similar to how None is an instance.

> I miss this feature too. We mention a couple of tools on https://llvm.org/docs/GitHub.html#using-graphite-for-stacked-pull-requests but I've not tried any of them myself.

Thanks, I'll give that a try for the next one

DavidSpickett reviewed (commented):
Some test is probably going to fail, but that's what we have buildbots for so I didn't look too closely at those changes.

Please fix the formatting and update the PR description to reflect the latest changes.

And I assume all the annotations are 3.8 compatible? At least according to vermin or pyright, whichever you prefer.

dsandersllvm committed:
fixup: type aliases is 3.12, 3.8 equivalent just drops the `type` keyword
